src/fu-util.c

```diff
@@ -3676,6 +3676,119 @@ fu_util_show_plugin_warnings(FuUtilPrivate *priv)
+		if (dependency_kind != NULL)
+			*dependency_kind = FU_UTIL_DEPENDENCY_KIND_RUNTIME;
+		return g_strndup(str + 15, strsz - 16);
+	}
+	if (g_str_has_prefix(str, "CompileVersion(")) {
+		if (dependency_kind != NULL)
+			*dependency_kind = FU_UTIL_DEPENDENCY_KIND_COMPILE;
+		return g_strndup(str + 15, strsz - 16);
+	}
+	return g_strdup(str);
+}
+
+static gboolean
+fu_util_version_as_json(FuUtilPrivate *priv, GHashTableIter *iter, GError **error)
+{
+	const gchar *key;
+	const gchar *value;
+	g_autoptr(JsonBuilder) builder = json_builder_new();
+
+	json_builder_begin_object(builder);
+	json_builder_set_member_name(builder, "Versions");
+	json_builder_begin_array(builder);
+	while (g_hash_table_iter_next(iter, (gpointer *)&key, (gpointer *)&value)) {
+		FuUtilDependencyKind dependency_kind = FU_UTIL_DEPENDENCY_KIND_UNKNOWN;
+		g_autofree gchar *project = fu_util_parse_project_dependency(key, &dependency_kind);
+		json_builder_begin_object(builder);
+		if (dependency_kind != FU_UTIL_DEPENDENCY_KIND_UNKNOWN) {
+			json_builder_set_member_name(builder, "Type");
+			json_builder_add_string_value(
+			    builder,
+			    fu_util_dependency_kind_to_string(dependency_kind));
+		}
+		json_builder_set_member_name(builder, "AppstreamId");
+		json_builder_add_string_value(builder, project);
+		json_builder_set_member_name(builder, "Version");
+		json_builder_add_string_value(builder, value);
+		json_builder_end_object(builder);
+	}
+	json_builder_end_array(builder);
+	json_builder_end_object(builder);
+	return fu_util_print_builder(builder, error);
+}
+
+static gboolean
+fu_util_version(FuUtilPrivate *priv, GError **error)
+{
+	const gchar *key;
+	const gchar *value;
+	GHashTableIter iter;
+	g_autoptr(GHashTable) metadata = NULL;
+	g_autoptr(GString) str = g_string_new(NULL);
+
+	/* get metadata */
+	metadata = fwupd_client_get_report_metadata(priv->client, priv->cancellable, error);
+	if (metadata == NULL)
+		return FALSE;
+	g_hash_table_iter_init(&iter, metadata);
+
+	/* remove anything that's not a version */
+	while (g_hash_table_iter_next(&iter, (gpointer *)&key, (gpointer *)&value)) {
+		if (!g_str_has_prefix(key, "RuntimeVersion") &&
+		    !g_str_has_prefix(key, "CompileVersion"))
+			g_hash_table_iter_remove(&iter);
+	}
+
+	/* dump to the screen in the most appropriate format */
+	g_hash_table_iter_init(&iter, metadata);
+	if (priv->as_json)
+		return fu_util_version_as_json(priv, &iter, error);
+	while (g_hash_table_iter_next(&iter, (gpointer *)&key, (gpointer *)&value)) {
+		FuUtilDependencyKind dependency_kind = FU_UTIL_DEPENDENCY_KIND_UNKNOWN;
+		g_autofree gchar *project = fu_util_parse_project_dependency(key, &dependency_kind);
+		g_string_append_printf(str,
+				       "%-10s%-30s%s\n",
+				       fu_util_dependency_kind_to_string(dependency_kind),
+				       project,
+				       value);
+	}
+
+	/* add this as a sanity check */
+	if (fwupd_client_get_host_bkc(priv->client) != NULL) {
+		g_string_append_printf(
```

Comment:
isn't this missed in the json?

Reply:
Yup, I just noticed that too :) I actually think adding it to the report metadata is a good idea too, so we can understand "odd" upgrade paths.